apps/labs/posts/libsf-error-state.md

```diff
@@ -0,0 +1,797 @@
+modules. Irwin and I begin working on this independently, comparing notes as we went.
+
+The first hiccup is that the following invocations were working on Irwin's Mac.
+
+Setting up the shared library like this.
+
+```python
+sf_error_state_lib = shared_library('sf_error_state', # Name of the library
+  # Implementation files contained it.
+  ['sf_error_state.c'],
+  # Additional places the preprocessor can look for header files
+  include_directories: ['../_lib', '../_build_utils/src'],
+  install: true
+)
+```
+
+and adding a `link_with` entry to the creation of each extension module:
+
+```python
+py3.extension_module('_special_ufuncs',
+  ['_special_ufuncs.cpp', '_special_ufuncs_docs.cpp', 'sf_error.cc'],
+  include_directories: ['../_lib', '../_build_utils/src'],
+  dependencies: [np_dep],
+  link_with: [sf_error_state_lib], # The new line
+  link_args: version_link_args,
+  install: true,
+  cpp_args: ['-DSP_SPECFUN_ERROR'],
+  subdir: 'scipy/special'
+)
+```
+
+However, on my Linux machine SciPy would build without issue, but attempting to import `scipy.special` would
+result in the error:
+
+```
+ImportError: libsf_error_state.so: cannot open shared object file: No such file or directory
+```
+
+After a period of head scratching in which I pondered every possible explanation except the correct one, I showed Irwin
+what I tried and the error message I was getting. It turned out that the difference in operating systems was a red
+herring. The issue was that, of the two methods for building SciPy from source for development recommended in SciPy's
+[contributor
+documentation](https://docs.scipy.org/doc/scipy/building/index.html#building-from-source-for-scipy-development), I was
+using the `"python dev.py build"` workflow and he was using an editable install: `"pip install -e . --no-build-isolation"`.
+For the editable install, SciPy is installed directly in its own project folder, and the shared library and the relevant
+extension modules were all being installed next to each other in `~/scipy/scipy/special`. For the `dev.py` workflow,
+SciPy is installed elsewhere. Since I didn't specify where to install the shared library, it got put in the wrong
+place. I fixed things by directly configuring the `install_dir` in Meson like this:
+
+```python
+sf_error_state_lib = shared_library('sf_error_state', # Name of the library
+  # Implementation files contained it.
+  ['sf_error_state.c'],
+  # Additional places the preprocessor can look for header files
+  include_directories: ['../_lib', '../_build_utils/src'],
+  install: true
+  # Tell meson where to install the shared library
+  install_dir: py3.get_install_dir() / 'scipy/special',
+)
+```
+
+Even after this, I was still seeing the same error. I asked Ralf about it, and found what happens is that `pip`
+goes through [meson-python](https://github.com/mesonbuild/meson-python) and its editable install support uses
```

Comment:
```suggestion
goes through [meson-python](https://github.com/mesonbuild/meson-python) and `pip`'s editable install support uses
```

The sentence is ambiguous by structure, and although I think context makes it clear that it's `pip` that has editable install support, not all readers might be aware of this. My preference is to be explicit to avoid the cognitive overhead of resolving the ambiguity.